File: A3/tli.py
#! /usr/bin/env python3
"""
//assume in the command prompt the user has entered the argument, tl.txt, the name of
//the text file that contains a TL program in it
create sList, an empty list of Stmt objects
create symTable, an empty symbol table
for each line of statement in tl.txt:
    parse the line to an Stmt object //possibly with satellite Expr objects
    add the Stmt object to sList
    if this line is labeled:
        add (label, current line number) mapping to symTable
evaluate sList with symTable
"""
import fileinput
import sys

def is_number(s):
    try:
        float(s)
        return True
    except ValueError:
        return False

# used to store a parsed TL expressions which are
# constant numbers, constant strings, variable names, and binary expressions
class Expr :
    def __init__(self, op1, operator, op2=None):
        self.op1 = op1
        self.operator = operator
        self.op2 = op2

    def __str__(self):
        if self.op2 == None:
            return self.operator + " " + self.op1
        else:
            return self.op1 + " " + self.operator + " " + self.op2

    # evaluate this expression given the environment of the symTable
    def eval(self, symTable):
        op1 = self.op1
        op2 = self.op2

        if is_number(str(op1)):
            symTable[op1] = float(op1)
        if is_number(str(op2)):
            symTable[op2] = float(op2)

        if self.operator == "var":
            return symTable[op1]
        elif self.operator == "string":
            op1 = str(op1.strip('"'))
            return op1
        elif self.operator == "constant":
            return op1
        elif self.operator == "+":
            return symTable[op1] + symTable[op2]
        elif self.operator == "-":
            return symTable[op1] - symTable[op2]
        elif self.operator == "*":
            return symTable[op1] * symTable[op2]
        elif self.operator == "/":
            return symTable[op1] / symTable[op2]
        elif self.operator == "<":
            return symTable[op1] < symTable[op2]
        elif self.operator == ">":
            return symTable[op1] > symTable[op2]
        elif self.operator == "<=":
            return symTable[op1] <= symTable[op2]
        elif self.operator == ">=":
            return symTable[op1] >= symTable[op2]
        elif self.operator == "==":
            return symTable[op1] == symTable[op2]
        elif self.operator == "!=":
            return symTable[op1] != symTable[op2]
        else:
            print('Syntax error on line {}.'.format(symTable["sListLineCount"] + 1))
            return sys.exit()

# used to store a parsed TL statement
class Stmt :
    def __init__(self,keyword,exprs):
        self.keyword = keyword
        self.exprs = exprs

    def __str__(self):
        others = ""
        for exp in self.exprs:
            others = others + " " + str(exp)
        return self.keyword + others

    # perform/execute this statement given the environment of the symTable
    def perform(self, symTable):
        exprs = self.exprs
        if self.keyword == 'let':
            tempExprs = ""
            if len(exprs) > 3:
                tempExprs = exprs[2]
                internal = Expr(exprs[2], exprs[3], exprs[4]).eval(symTable)
                exprs[2] = str(internal)

            if str(exprs[2]).isdigit():
                symTable[exprs[0]] = float(exprs[2])
            else:
                while not is_number(str(exprs[2])):
                    exprs[2] = symTable[exprs[2]]

                if is_number(str(exprs[2])):
                    symTable[exprs[0]] = float(exprs[2])
            if len(tempExprs) > 0:
                exprs[2] = tempExprs
        # Print
        elif self.keyword == 'print':
            finalPrint = ""
            for index, expressions in enumerate(exprs):
                expressionsList = expressions.split()
                if len(expressionsList) == 3:
                    evaluatedExpression = Expr(expressionsList[0],expressionsList[1],
                                               expressionsList[2]).eval(symTable)
                    symTable[expressions] = str(evaluatedExpression)

            for index,expressions in enumerate(exprs):
                if exprs[index] in symTable:
                    evaluatedExpression = Expr(exprs[index],'var').eval(symTable)
                elif exprs[index] not in symTable and '"' in exprs[index]:
                    evaluatedExpression = Expr(exprs[index],'string').eval(symTable)
                elif exprs[index] not in symTable and '"' not in exprs[index]:
                    evaluatedExpression = Expr(exprs[index],'constant').eval(symTable)
                finalPrint += str(evaluatedExpression) + " "

            print(finalPrint[:-1])
        # Input
        elif self.keyword == 'input':
            for expressions in exprs:
                newInput = input()
                if str(newInput).isdigit():
                    newInput = float(newInput)
                    symTable[expressions] = newInput
                else:
                    sys.exit('Illegal or missing input')
        # Goto:
        # While reading in strings into statements during main, we check if a word is a label
        # If it is, then the word is added into the symbol table along with the line number
        # As we perform if, we know that the first 3 expressions will be the actual expression
        # that we need to evaluate to see if the if statement is true or false. Since our third
        # expression is always 'goto' we can just ignore it with our evaluations.
        #
        # So, line by line this is what we perform:
        elif self.keyword == 'if':
            # take the label that we will be jumping to
            label = exprs[4]
            # evaluate the expression, if true, then we change which line we should be on using
            # our value "sListLineCount" that we saved into the symbol table ( this tells us
            # which statement line we are currently evaluating during perform). If our expression
            #  is true then this line number changes to the labeled line, otherwise,
            # it just iterates up one and continues to the next line
            if Expr(exprs[0], exprs[1], exprs[2]).eval(symTable):
                symTable["sListLineCount"] = int(symTable.get(label))
            else:
                symTable["sListLineCount"] = int(symTable["sListLineCount"] + 1)

        # since we are using "sListLineCount" to actually read which statements we are about to
        # perform, if we haven't used an if statement, then we just iterate up one
        if self.keyword != 'if':
            symTable["sListLineCount"] = int(symTable["sListLineCount"] + 1)


def main():

    symTable = {}
    symTable["sList"] = []
    # "sListLineCount" keeps track of which index in our list of statements we are currently
    # performing.
    # Ex: if 5 statements, sListLineCount will start at 0, then finish at 4
    symTable["sListLineCount"] = 0
    sList = symTable["sList"]

    # just a counter so we can see what line we're on when we throw errors
    lineCount = 0

    # Takes in our .txt file, and for each line in the file, we run the for loop:
    for line in fileinput.input(sys.argv[1:]):

        # lines.strip() takes the line as a string and throws it into a list called "lines"
        lines = line.strip()

        # takes the lines and turns them into a list of each string
        # ie let x = 5 becomes ['let', 'x', '=', '5']
        stringList = lines.split()

        if ':' in lines:
            symTable[stringList[0].strip(":")] = lineCount
            stringList.remove(stringList[0])

        # sets keyword to the first item in our stringList list
        keyword = stringList[0]

        # creates an empty array that we will fill in depending on the length of the expression
        expression = []
        # creates an array of legal keywords
        legalKeywords = ["let", "if", "print", "input"]

        # check to make sure keyword is spelled correctly
        if keyword not in legalKeywords:
            print('Syntax error on line {}.'.format(lineCount + 1))
            sys.exit()

        # now that we know the keyword, we get rid of it
        stringList.remove(stringList[0])

        # check if its a print line, if it is, adjust so prints properly\
        if keyword == "print":
            printExprs = ""
            for strings in stringList:
                printExprs += strings + " "

            printExprs = printExprs[:-1]
            printExprs = printExprs.split(' , ')
            for expressions in printExprs:
                expression.append(expressions)
        else:
            # now we add the rest of the strings into expression
            for strings in stringList:
                expression.append(strings)

        # our statement has been built, so now we add it to the list of statements
        statement = Stmt(keyword, expression)
        sList.append(statement)

        lineCount += 1

    count = 0
    while count < len(symTable["sList"]):

        sListLineCount = symTable["sListLineCount"]
        sList[sListLineCount].perform(symTable)
        count = int(sListLineCount) + 1


if __name__ == "__main__":
    main()

File: A4/TLI.scala
import scala.collection.mutable.Map
import scala.collection.mutable.ArrayBuffer
import scala.io.Source

// Create a class called Expr
class Expr(op1: Any, operator: Any, op2: Any){
	override def toString = op1 + " " + operator + " " + op2
	
	def eval(symTable: Map[Any,Any]) :Any = {
		
		if (operator == "var"){
			val result = symTable(op1)
			return result
		} else if (operator == "+"){
			val first = symTable(op1).asInstanceOf[Double]
			val second = symTable(op2).asInstanceOf[Double]
			val result = first + second
			return result
		}
	}
}
//hide given case classes
//case class Var(name: String) extends Expr
//case class Str(name: String) extends Expr
//case class Constant(num: Double) extends Expr
//case class BinOp(operator: String, left: Expr, right: Expr) extends Expr

// Create a class called Stmt
class Stmt(keyword: String, exprs: Array[String]){

	def copy(newKeyword: String = keyword, newExprs: Array[String] = exprs): Stmt =
		new Stmt(newKeyword, newExprs)
		
	var others = ""
	exprs.foreach(exp => others = others + " " + exp.mkString(" "))
	
	override def toString = keyword + others
	
	def perform(symTable: Map[Any,Any]){
		
		println("keyword: " + keyword )
		println("exprs: " + exprs.mkString(" "))
		
		if (keyword == "let"){
			var tempExprs = ""
			var test = exprs(2).toDouble
			symTable.addOne(exprs(0) -> test)
		}else if (keyword == "print"){
			var finalPrint = ""
			var evaluatedExpression = new Expr(exprs(0),"var",None).eval(symTable)
			finalPrint = finalPrint + evaluatedExpression.toString()
			
			println(finalPrint)
			
		}
		var updateCount = symTable("sListLineCount").asInstanceOf[Integer]
		updateCount += 1
		symTable.update("sListLineCount", updateCount)
	}
}
//hide given caseclasses
//case class Let(variable: String, expr: Expr) extends Stmt
//case class If(expr: Expr, label: String) extends Stmt
//case class Input(variable: String) extends Stmt
//case class Print(exprList: List[Expr]) extends Stmt

object TLI {
	//hiding given eval
    //def eval(expr: Expr, symTab: Map[String, Double]): Double = expr match {
    //    case BinOp("+",e1,e2) => eval(e1,symTab) + eval(e2,symTab) 
    //    case Var(name) => symTab(name)
    //    case Constant(num) => num
	//case _ => 0 // should really throw an error
    //}

    def main(args: Array[String]) {
    	// symbolTable dictionary
    	var symTable:Map[Any,Any] = Map()
    	
    	// sList  is our list of Statements
    	var sList = Array[Stmt]()
    	
    	// just a counter so we can see what line we're on when we throw errors
    	var lineCount = 0
    	
    	// keeps track of which index in our list of statements we are currently
    	// performing.
    	symTable.addOne("sListLineCount" -> 0)
    	
    	// Takes in our .txt file, and for each line in the file, we run the for loop:
    	var filename = "test.txt"
    	
		for (line <- Source.fromFile(filename).getLines) {
    		// read in each line and remove all blank spaces and drop in ","
    		var afterLine = line.replace(" ", ",")
    		
    		// create a an empty list
    		var stringList = Array[String]()
    		
    		// add in the strings from afterLine but separating elements whenver it sees ","
    		stringList = afterLine.split(",").map(_.toString).distinct
    		
    		// set keyword to the first element in stringList
    		var keyword = stringList(0)
    		
    		// create an array called expression
    		var expression = Array[String]()
    		
    		// create an array called legalKeywords, they have all valid keywords
    		var legalKeywords = Array("let", "if", "print", "input")
    		
    		// create a boolean of whether or not keyword is in the array of keywords
    		val res = legalKeywords.contains(keyword)
    		
    		// if boolean res does not equal true, print an error
    		if (res == false) {
    			println("Syntax error on line " + (lineCount + 1) + ".")
    			System.exit(0)
    		}

			// remove keyword from stringList and set it to a new array called stringLists
    		var stringLists = stringList.filter(! _.contains(keyword))
    		println(stringLists.mkString(" "))
    		
    		//if (keyword == "print") {
    		//	var printExprs = ""
    		//	stringLists.foreach(strings => printExprs += strings + " " )
    		//	
    		//	var printExprsAfter = printExprs.split(",").map(_.toString).distinct
    		//	
    		//	println("so exprs is: ")
    		//	println(printExprsAfter.mkString(" "))
    		//}
    		
    		stringLists.foreach(strings => expression = expression:+ strings)
    		
    		println("expressions are: " + expression.mkString(" "))
    		
    		println("expression: " + expression.mkString(" "))
			println("keyword: " + keyword)
			var statement = new Stmt(keyword, expression)
			
			sList = sList:+ statement
			
    		lineCount += 1
		}
    	symTable.addOne("sList" -> sList)
    	var count = 0
    	var sListLen = sList.size
    	
    	
    	while (count < sListLen){
    		var sListLineCount = symTable("sListLineCount").asInstanceOf[Integer]
    		sList(sListLineCount).perform(symTable)
    		count = sListLineCount + 1
    	}
    }
}
